### DISCUSSION TITLE
Allow tabs to be renamed

### DISCUSSION THREAD

**Discussion by mr-cheffy** [zen-browser/desktop#869] [Ideas]
### What feature would you like to see?

I’d love to be able to rename tabs in the tabs bar. This is a killer feature of Arc for me. As a web developer, I’ll often have multiple versions of the same page open (i.e. different environments) and being able to differentiate them is very helpful.

### What platform are you seeing the problem on?

macOS

 
 ~ _Created automatically from issue: https://github.com/zen-browser/desktop/issues/261_

**Comment by ydivho12123** [comment]
I really need this feature, both ARC and Vivaldi have this rename tab feature, this is the only reason I still have Vivaldi browser, if there is this feature then I can finally ditch Vivaldi completely and use Zen completely. This feature can greatly reduce the stress of distinguishing tabs. Although there are related plug-ins, but those plug-ins are very rudimentary and unstable, often refresh or restart the browser after the renaming is invalid, or can only rename the entire domain name, and the link only needs to change slightly, for example, the video website just automatically plays the next recommended video, the link changes, the renaming is invalid, basically completely unavailable, this function can only be achieved by the browser internally

**Comment by FrancisMorelli** [comment]
Please, consider prioritize and develop this feature. This is one of the best Arc's native features.

**Comment by negativems** [comment]
Need it!

**Comment by SamBozman** [comment]
Right now the tabs show the current url of the open web site and this is confusing. I would like to have the tab show the name of the link, not the url please. If this web browser can fix most of the outstanding issues, it will be an outstanding browser! :)

**Reply by rubaboo** [reply]
Is it not like this for you? The only place I see the URL is in the URL bar.

### Tab shows a title
![image](https://github.com/user-attachments/assets/ca330270-a453-4f3c-96d9-f09a17171f97)

### Taskbar shows a title as well
![image](https://github.com/user-attachments/assets/1af444fe-5aa7-439d-aa79-39b7c8913630)

**Comment by Jrzpvt** [comment]
Really wish this feature existed, by a Windows user.

**Comment by Muhammad-Mohid-Khan** [comment]
i want this feature as well , as a windows user

**Comment by kvenn** [comment]
I also would love this feature. When working with multiple pages that have the same name (like reviewing analytics charts, for example), it's so helpful to rename them based on the chart.

A bonus request would be: when opening a bookmark which has a custom title - apply that title to the newly opened tab. 
Bonus BONUS request would be: allowing single-click on the bookmark to jump to the currently open tab (if already open) and cmd+click to always open it in a new tab

**Comment by rubaboo** [comment]
You can do it now. No reason to wait until Zen is as bloated as the previous favorite browser that is now dead.

https://addons.mozilla.org/en-CA/firefox/search/?q=tab%20title
https://addons.mozilla.org/en-CA/firefox/search/?q=rename%20tab

**Reply by kvenn** [reply]
I would personally prefer if it was supported in the browser (as opposed to an extension). I tried the extension for a few weeks but it was a couple steps back from just being able to rename either 1) through the command palette or 2) through clicking on the tab

**Reply by vraravam** [reply]
sorry to take a tangential question: is there a command pallette?

**Reply by kvenn** [reply]
CMD+E by default. But I added CMD+P as well (since....who prints things?)

**Reply by rubaboo** [reply]
> CMD+E by default. But I added CMD+P as well (since....who prints things?)

I don't think that is what people usually understand as a "command palette". That is just a search box. I see no reason to use that, it's less functional than Alt+D (focus URL bar).

**Reply by kvenn** [reply]
oops that's my bad, I thought I was on the Vivaldi thread. You're totally right, command palette does not exist in Zen. I'm currently trying out all the browsers on my shift off of Arc and have filed lots of complaints on both (ha)

**Reply by Waquo** [reply]
Unfortunately these extensions do not work for pdf tabs, which is the one situation where I really want to be able to rename tabs.
A tab named 2412.21187v1.pdf does not tell me which arxiv paper it is.

**Reply by aidoanto** [reply]
Those extensions don't work for me, (some don't work at all and others eventually stop using the tab title you've given it and go back to the old one. it would seem there might be some difference between the functionality on vanilla firefox and zen, but I'm not sure.

**Comment by Regen-forest** [comment]
This is very important, as mentioned above.

**Reply by rubaboo** [reply]
Sorry, but it's not. There's plenty of other bugs and visual glitches that should be fixed first, before introducing yet another feature that former Arc users must have.

**Reply by jeanlucthumm** [reply]
@rubaboo you're condescending to Arc users throughout your comments. It's not constructive, stop.

**Reply by syscat29** [reply]
> Sorry, but it's not. There's plenty of other bugs and visual glitches that should be fixed first, before introducing yet another feature that former Arc users must have.

Please my guy keep it civil, if you're not going to offer any constructive feedback refrain from commenting.

Do not make me tap the button.

**Comment by afrianluthfan** [comment]
Yes please. This is literally the only thing holding me back from switching to Zen from Arc.

**Reply by 12th-devs** [reply]
There is a way but it is a little complex...

https://github.com/wysh3/zen-premium

This gives you the ability to rename tabs by double clicking them, or when you hover (if you want) it will show a little circle that will have an AI summarize it for you. (I use both of them and love it)

**Comment by gillpeter** [comment]
Yes, please add a tab rename feature. Its a pretty simple thing for a developer to enable it please add it to your queue @j14i

**Comment by 117649** [comment]
Tab mix plus is working on support zen. https://github.com/onemen/TabMixPlus/issues/371#event-16067637923
You can try.

**Comment by issuex** [comment]
to dev, why not just add a special folder to bookmark and use it for store tab data that user can also rename tab or add folder or sync it, make ui to combine tab and the bookmark that store in special folder to make it look like using normal tab.

for example.
/ Bookmark Menu / ZEN-TAB / WORKSPACE-1 / \<all tab here, store as bookmark \>
/ Bookmark Menu / ZEN-TAB / WORKSPACE-2 /\< user dir \>/ \<all tab here, store as bookmark \>

**Comment by 12th-devs** [comment]
https://github.com/wysh3/zen-premium

**Comment by joaolisboa** [comment]
Would love to see this built-in to the browser without needing mods or extensions

**Comment by Rexxyyy** [comment]
how is this not a thing yet? Please push this before stable.

**Comment by 12th-devs** [comment]
Good News!!!

https://github.com/zen-browser/desktop/pull/4979

**Comment by lockieluke** [comment]
This is what I'm using to rename any tab by pressing F2, not ideal though

```javascript
// ==UserScript==
// @name        Rename Tab
// @namespace   Violentmonkey Scripts
// @match       *
// @grant       none
// @version     1.0
// @author      -
// @description 24/02/2025, 10:39:10
// ==/UserScript==

window.addEventListener("keydown", event => {
    const renaming = event.key === "F2";
    if (renaming)
        document.title = prompt("Rename Tab");
});
```

**Reply by roman-balzer** [reply]
"not ideal" you say. I think it's genius and so simple

**Reply by roman-balzer** [reply]
I had to change `@match` from `*` to `*://*/*` to make it work

**Comment by niderhoff** [comment]
it's part of 1.8b release

**Reply by gillpeter** [reply]
But it only allows for pinned tabs to be renamed. What about other tabs?

**Reply by sahabji0P** [reply]
Currently it does not allows for the pinned tabs also in the Zen browser

**Reply by rajesh-zero** [reply]
> Currently it does not allows for the pinned tabs also in the Zen browser

it does
This feature works actually just have to double click on the pinned tabs to rename

**Comment by sahabji0P** [comment]
I would also like the feature to add tabs to the folders (Pinned tabs) for cleaner workspace (a good implementation done by ARC)

**Reply by rajesh-zero** [reply]
This feature works actually just have to double click on the pinned tabs to rename

**Comment by rajesh-zero** [comment]
Yes please I was trying to move n completely from arc and started to set up.
But not having this feature is stopping me.

I know this is free and open source and we should not rant but i really wish i could.

I am also a software developer. I tried to look at code to contribute, but I feel very overwhelmed.

**Reply by rajesh-zero** [reply]
This feature works actually just have to double click on the pinned tabs to rename

**Reply by rajesh-zero** [reply]
It would be nice if this was seen in right click context menu

**Comment by FrancisMorelli** [comment]
Great job. It would be nice if this was seen in right click context menu.

**Comment by roman-balzer** [comment]
Also would be great if it would'nt be overridden by the site :/

**Reply by ChristinWhite** [reply]
I think that's the key issue with all the current solutions, the title needs to be persistent.

**Comment by mr-cheffy** [comment]
I think this should be closed now, as this feature has existed for quite some time!

**Reply by nithinbull** [reply]
The feature as of now seems incomplete. It renames the tab visually, but when searched through the cmt+t search bar, the renamed tabs don't appear with their new names, rather they're shown with their original page title. This is inconsistent.

**Reply by nithinbull** [reply]
[arc_feature_parity_on_pinned_tabs](https://github.com/zen-browser/desktop/discussions/2829#discussioncomment-12593240)
Please refer the above comment for further context.

**Comment by FrancisMorelli** [comment]
I think this topic has not been fully implemented, because if click on a different link within the pin tab website, the name changes. It should stay the way you chose.

**Comment by lockieluke** [comment]
If the condition check on `pinned` is removed here, we will be able to rename unpinned tabs
https://github.com/zen-browser/desktop/blob/6cdc6409772850f1424a71c4a8f5fae7360dc5ef/src/zen/common/ZenUIManager.mjs#L1074

**Reply by NikilXYZ** [reply]
+1 yep please remove this unnecessary check for pinned tabs

**Reply by melecoaze** [reply]
+1 this feature would improve my current workflow by miles

**Reply by alexandrelissowski** [reply]
Why don't you suggest a PR?

**Comment by KParthSingh** [comment]
When you right click on a tab there should be an option in the context menu to rename tab

**Comment by omidiw** [comment]
hey i really need the rename tab option as a window user as well. arc got it and it's the only feature left from arc that i don't have on zen browser. i wanna ditch arc totally please help in process.

**Reply by watjurk** [reply]
It is done. Double click on a tab

**Reply by kvenn** [reply]
I believe you can only rename pinned tabs. In Arc (and others, like Vivaldi) you can rename any tab.

**Reply by azeinier** [reply]
&gt; It is done. Double click on a tab

Thanks ! It works. That's all I was missing on Zen compared to Arc.

**Reply by hypnotie** [reply]
I would love to have a possibility to rename any tab, not just the pinned ones :(

**Comment by alexandrelissowski** [comment]
I would like that Zen automatically remove the dynamic numbers from the title. 
Some websites, like Notion use the title to display the number of pending notifications. As I don't use Notion notifications, I feel it very annoying to always see `(+9)` in the tab name. 

Example: 
<img width="568" height="110" alt="Screenshot 2025-09-02 at 20 05 28" src="https://github.com/user-attachments/assets/4ee6b476-49fa-483e-948b-6e14674a7e53" />
